### PR TITLE
Use shared secret and receipt field

### PR DESCRIPTION
**Pull request by KLVTZ**
Description
------------
In addition to what was adopted from existing PRs. we added additional features. This includes providing defaults in case receipt field values are not defined. We also provide the client the ability to access latest receipt information when available. This is important when validating up-to-date monthly accounts that auto-renew. By default we use the receipt's in_app as a a starting point.

Goal is to maintain backwards compatibility with different kinds of apple receipts while providing default values.

Test
----
- `./verify-apple.js --productId=<optional> --packageName=<optional> --receipt='<payload>' -- secret='<password>'
- https://gist.github.com/KLVTZ/504d01ef3db1d05d9e43b5199f110789

Resolves
---------
#23 - Adopts PR a starting point in using the shared secret for requests -- merged into master.
#11 - Adopts PR to support receipt field retrieval

### CODE REVIEW

**Comment by ronkorving** [comment]
Could you pull master and reduce this PR to the relevant changes?

**Comment by KLVTZ** [comment]
@ronkorving updated with master.

File: lib/apple/index.js [hunk 1]
```diff
@@ -32,14 +46,27 @@ function parseResult(result) {
 		throw error;
 	}
 
-	var receipt = result.receipt;
+	var productId = getReceiptFieldValue(result.receipt, 'product_id');
+	var transactionId = getReceiptFieldValue(result.receipt, 'transaction_id');
+
+	/* jshint camelcase:false */
+    if (result.hasOwnProperty('latest_receipt_info') && result.latest_receipt_info[0]) {
+        latestReceiptInfo = result.latest_receipt_info.sort(function (a, b) {
```
Comment (ronkorving) [inline lib/apple/index.js]: Indentation's a bit funky

File: lib/apple/index.js [hunk 2]
```diff
@@ -32,14 +46,27 @@ function parseResult(result) {
...
+            a = parseInt(a.original_purchase_date_ms);
+            b = parseInt(b.original_purchase_date_ms);
```
Comment (ronkorving) [inline lib/apple/index.js]: Sorry, this is my bad for not hard-enforcing a style on this project, but would you mind adding a radix argument (`parseInt(str, 10)`) on these two lines?

File: lib/apple/index.js [hunk 3]
```diff
@@ -32,14 +46,27 @@ function parseResult(result) {
...
+
+            return a - b;
+        });
+
+        productId = latestReceiptInfo.slice(-1).pop().product_id;
+        transactionId = latestReceiptInfo.slice(-1).pop().transaction_id;
```
Comment (ronkorving) [inline lib/apple/index.js]: Sorry, I'm a bit sensitive to performance nits. Would you mind reading from `latestReceiptInfo[latestReceiptInfo.length - 1]` instead?

File: lib/apple/index.js [hunk 4]
```diff
@@ -17,12 +17,26 @@ var responses = {
...
+	if (receipt.hasOwnProperty(field)) {
+		return receipt[field];
+	}
+
+	/* jshint camelcase:false*/
+	if (receipt.hasOwnProperty('in_app') && receipt.in_app[0] && receipt.in_app[0].hasOwnProperty(field)) {
+		return receipt.in_app[0][field];
+	}
+
+    return null;
```
Comment (ronkorving) [inline lib/apple/index.js]: Indentation issue

File: lib/apple/index.js [hunk 5]
```diff
@@ -100,18 +127,26 @@ exports.verifyPayment = function (payment, cb) {
...
+		
+		var productId = getReceiptFieldValue(receipt, 'product_id');
+
+		if (payment.hasOwnProperty('productId') && payment.productId !== productId) {
+			return cb(new Error('Wrong product ID: ' + payment.productId + ' (expected: ' + productId + ')'));
+		}
+		
+		var receiptBundleId = getReceiptFieldValue(receipt, 'bid');
 
 		/* jshint camelcase:false */
```
Comment (ronkorving) [inline lib/apple/index.js]: No need for this jshint annotation anymore.

File: lib/apple/index.js [hunk 6]
```diff
@@ -100,18 +127,26 @@ exports.verifyPayment = function (payment, cb) {
...
-		if (payment.hasOwnProperty('productId') && payment.productId !== receipt.product_id) {
-			return cb(new Error('Wrong product ID: ' + payment.productId + ' (expected: ' + receipt.product_id + ')'));
+		if (receiptBundleId === null) {
+			receiptBundleId = getReceiptFieldValue(receipt, 'bundle_id');
 		}
 		/* jshint camelcase:true */
```
Comment (ronkorving) [inline lib/apple/index.js]: (and so this one too)

**Comment by ronkorving** [comment]
Could you also update the readme to document the new `latestReceiptInfo` field?

File: lib/apple/index.js (hunk 1 above)
Comment (KLVTZ) [inline lib/apple/index.js]: Fixed.

File: lib/apple/index.js (hunk 2 above)
Comment (KLVTZ) [inline lib/apple/index.js]: No worries, your helping me to develop good practices --appreciate it! Plus it doesn't hurt to follow Mozilla via "Always specify this parameter to eliminate reader confusion and to guarantee predictable behavior."

File: lib/apple/index.js (hunk 3 above)
Comment (KLVTZ) [inline lib/apple/index.js]: Fixed.

File: lib/apple/index.js (hunk 4 above)
Comment (KLVTZ) [inline lib/apple/index.js]: Fixed.

File: lib/apple/index.js (hunk 5 above)
Comment (KLVTZ) [inline lib/apple/index.js]: Removed.

File: lib/apple/index.js (hunk 6 above)
Comment (KLVTZ) [inline lib/apple/index.js]: Removed

**Comment by KLVTZ** [comment]
@ronkorving resolved issues highlighted as well as added a sentence referencing when `in_app` and/or `latestReceiptInfo` may be available.

File: README.md [hunk 7]
```diff
@@ -59,35 +53,74 @@ contains your In-App Purchase Shared Secret
...
+						"original_purchase_date": "2016-10-29 22:46:57 Etc/GMT",
+						"original_purchase_date_ms": "1477781217000",
+						"original_purchase_date_pst": "2016-10-29 15:46:57 America/Los_Angeles",
+						"expires_date": "2016-11-30 05:46:42 Etc/GMT",
+						"expires_date_ms": "1480484802000",
+						"expires_date_pst": "2016-11-29 21:46:42 America/Los_Angeles",
+						"web_order_line_item_id": "820817285",
+						"is_trial_period": "false"
+					},
+				]
```
Comment (ronkorving) [inline README.md]: This block is strangely indented

File: README.md [hunk 8]
```diff
@@ -59,35 +53,74 @@ contains your In-App Purchase Shared Secret
...
         },
-        "transactionId": "1000000102526671",
+		"latestReceiptInfo": [
```
Comment (ronkorving) [inline README.md]: This block too

**Comment by ronkorving** [comment]
Sorry, last indentation issue (readme this time). After that I'm good to go 👍

**Comment by KLVTZ** [comment]
@ronkorving no worries, thanks for pointing them out. Fixed the issues you highlighted.

**Comment by KLVTZ** [comment]
@ronkorving in addition to resolving the formatting, we noticed that it would be better to sort by transaction id versus original purchase date. This is to guarantee an ascending order of latest receipts available.

File: lib/apple/index.js [hunk 9]
```diff
@@ -32,18 +46,27 @@ function parseResult(result) {
...
 	}
 
-	var receipt = result.receipt;
+	var productId = getReceiptFieldValue(result.receipt, 'product_id');
+	var transactionId = getReceiptFieldValue(result.receipt, 'transaction_id');
+
+	/* jshint camelcase:false */
+	if (result.hasOwnProperty('latest_receipt_info') && result.latest_receipt_info[0]) {
+		latestReceiptInfo = result.latest_receipt_info.sort(function (a, b) {
+			return parseInt(a.transaction_id, 10)  > parseInt(b.transaction_id, 10);
```
Comment (ronkorving) [inline lib/apple/index.js]: `sort` expects a numerical return value, where you're returning a boolean.

You could however easily address by replacing `>` with `-`.

Comment (KLVTZ) [inline lib/apple/index.js]: @ronkorving resolved it

**Comment by KLVTZ** [comment]
@ronkorving provided numerical value instead of boolean for sort. How are we looking? :)

**Comment by ronkorving** [comment]
I'm sorry to keep this dragging on. Will try to merge this as soon as possible. I just had some remaining questions about all the properties and after trying to read through [this recent Apple support thread](https://forums.developer.apple.com/thread/69531), my head is hurting...

I suggest, for better or for worse, that you also have a read through (sorry, I'm not trying to make your head hurt). After that, let's quickly discuss if we should label the new property "experimental".

**Comment by KLVTZ** [comment]
@ronkorving wanted to let you know I've seen the comment above and will put aside some time next week to ready and reply back. Thanks for looking into this a bit more. Side Note: wouldn't mind nudging #25 forward in the meantime ;)

**Comment by kelchy** [comment]
what is the hold up over here, this pull request solves a broken apple receipt format. would benefit a lot to merge this. the conflict in README.md is minor IMO

**Comment by KLVTZ** [comment]
@kelchy sorry for the holdup. This is pending a requirement for me to read over 

https://forums.developer.apple.com/thread/69531

I need to provide @ronkorving some feedback in implementing this fix under an "experiment" labeled property. However, if you have examples of receipts where this PR correctly supports your purchases. Then I would argue that this implementation is the new standard and not just "experimental."

In fact, we are using this exact logic in handling renewable subscriptions and validating monthly entitlement for our product in production. It's out in the wild and hasn't cause us any conflicts. I believe the maintainer of this package would have more confidence if you provided your case for this too.

Thanks for reaching out!

**Comment by kelchy** [comment]
hi justin,

if it works for you and it works for me, then i guess we should push it.
better to have this than something which is not working at all. i would assume a lot of developers who has intentions of using node-iap and collaborating would go away once they realize apple validation does not even work.

**Comment by ronkorving** [comment]
Any chance someone can provide an encoded test receipt, so we can add this to a unit test later?

**Comment by KLVTZ** [comment]
@ronkorving Thanks for getting back to us! So I provided a receipt example in the readme. All I did was changed some data that was personal to the customer. I could find an expired receipt in our database and provide that for reference. Any thoughts on how you want this receipt? Just a gist?

**Comment by ronkorving** [comment]
@KLVTZ Thanks! You can paste it in a gist and link it in this thread if you like 👍

**Comment by KLVTZ** [comment]
@ronkorving straight from the PR to your living-room 😀 

https://gist.github.com/KLVTZ/139b11295f3844d24e8418a78cf01e37

**Comment by ronkorving** [comment]
Thanks 👍

**Comment by ronkorving** [comment]
It took a while, sorry all :) I'll make a release soon.

**Comment by ronkorving** [comment]
Pushed 0.7.0.

**Comment by kelchy** [comment]
hooooraaay! :D

**Pull request merged**